teuthology/provision.py

```diff
@@ -29,105 +30,200 @@ def _get_downburst_exec():
+    A class that provides methods for creating and destroying virtual machine
+    instances using downburst: https://github.com/ceph/downburst
     """
-    status_info = get_status(machine_name)
-    if not status_info.get('is_vm', False):
-        return False
-    phys_host = decanonicalize_hostname(status_info['vm_host']['name'])
-    os_type = get_distro(ctx)
-    os_version = get_distro_version(ctx)
+    def __init__(self, name, os_type, os_version, status=None):
+        self.name = name
+        self.os_type = os_type
+        self.os_version = os_version
+        self.status = status or get_status(self.name)
+        self.config_path = None
+        self.host = decanonicalize_hostname(self.status['vm_host']['name'])
+        self.executable = downburst_executable()
 
-    createMe = decanonicalize_hostname(machine_name)
-    with tempfile.NamedTemporaryFile() as tmp:
-        has_config = hasattr(ctx, 'config') and ctx.config is not None
-        if has_config and 'downburst' in ctx.config:
-            log.warning(
-                'Usage of a custom downburst config has been deprecated.'
-            )
+    def create(self):
+        """
+        Launch a virtual machine instance.
 
+        If creation fails because an instance with the specified name is
+        already running, first destroy it, then try again. This process will
+        repeat two more times, waiting 60s between tries, before giving up.
+        """
+        if not self.executable:
+            log.error("No downburst executable found.")
+            return False
+        self.build_config()
+        success = None
+        with safe_while(sleep=60, tries=3,
+                        action="downburst create") as proceed:
+            while proceed():
+                (returncode, stdout, stderr) = self._run_create()
+                if returncode == 0:
+                    log.info("Downburst created %s: %s" % (self.name,
+                                                           stdout.strip()))
+                    success = True
+                    break
+                elif stderr:
+                    # If the guest already exists first destroy then re-create:
+                    if 'exists' in stderr:
+                        success = False
+                        log.info("Guest files exist. Re-creating guest: %s" %
+                                 (self.name))
+                        self.destroy()
+                    else:
+                        success = False
+                        log.info("Downburst failed on %s: %s" % (
+                            self.name, stderr.strip()))
+                        break
+            return success
+
+    def _run_create(self):
+        """
+        Used by create(), this method is what actually calls downburst when
+        creating a virtual machine instance.
+        """
+        if not self.config_path:
+            raise ValueError("I need a config_path!")
```

Comment:
usually logs or error messages are better presented without using the first person in them, so in this case, it would look like:

    raise ValueError('a config_path is required')